generate_perror/main.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"sort"
+	"strconv"
+)
+
+const (
+	fileHeader = `// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Generated by generate_perror/main.go
+package main
+
+var MysqlErrNameToNum = map[string]int{
+`
+)
+
+var (
+	tidbCodePath string
+)
+
+func init() {
+	flag.StringVar(&tidbCodePath, "path", "../tidb", "Path to TiDB source code root directory.")
```

Comment:
Does this mean we need to rely on tidb?

Reply:
Yes, in the same way as it relies on `perror` (a binary from the mysql distribution, to map error numbers to error code names).
If TiDB would have a similar program/binary/utility, to map error codes to names, we could use that instead of reading the source code.

But since we are generating the perror.go file, with the mapping, and that is checked into the mysql-tester repository, it only depends on `perror` and `tidb` repository when it needs to be updated. So it does not have any direct dependencies, only needs the dependencies when new error codes needs to be added.

Reply:
Got it.